### PR TITLE
Store time in session as java.util.Date rather than joda-time

### LINKED ISSUE #27: No reader function for tag object when using cookie store
If I have in my ring configuration

```(assoc-in site-defaults [:session :store] (cookie-store {:key "a very secret cookie key"}))```

I get the error:

```
java.lang.RuntimeException: No reader function for tag object
	at clojure.lang.EdnReader$TaggedReader.readTagged(EdnReader.java:801)
	at clojure.lang.EdnReader$TaggedReader.invoke(EdnReader.java:783)
	at clojure.lang.EdnReader$DispatchReader.invoke(EdnReader.java:549)
	at clojure.lang.EdnReader.readDelimitedList(EdnReader.java:757)
	at clojure.lang.EdnReader$MapReader.invoke(EdnReader.java:680)
	at clojure.lang.EdnReader.readDelimitedList(EdnReader.java:757)
	at clojure.lang.EdnReader$MapReader.invoke(EdnReader.java:680)
	at clojure.lang.EdnReader.readDelimitedList(EdnReader.java:757)
	at clojure.lang.EdnReader$MapReader.invoke(EdnReader.java:680)
	at clojure.lang.EdnReader.read(EdnReader.java:145)
	at clojure.lang.EdnReader.read(EdnReader.java:111)
	at clojure.lang.EdnReader.readString(EdnReader.java:67)
	at clojure.edn$read_string.invokeStatic(edn.clj:46)
	at clojure.edn$read_string.invokeStatic(edn.clj:37)
	at clojure.edn$read_string.invoke(edn.clj:37)
	at ring.middleware.session.cookie$serialize.invokeStatic(cookie.clj:69)
	at ring.middleware.session.cookie$serialize.invoke(cookie.clj:68)
	at ring.middleware.session.cookie$seal.invokeStatic(cookie.clj:75)
	at ring.middleware.session.cookie$seal.invoke(cookie.clj:72)
	at ring.middleware.session.cookie.CookieStore.write_session(cookie.clj:91)
	at ring.middleware.session$bare_session_response.invokeStatic(session.clj:52)
	at ring.middleware.session$bare_session_response.invoke(session.clj:43)
	at ring.middleware.session$session_response.invokeStatic(session.clj:74)
	at ring.middleware.session$session_response.invoke(session.clj:66)
	at ring.middleware.session$wrap_session$fn__24588.invoke(session.clj:109)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__24630.invoke(keyword_params.clj:36)
	at ring.middleware.nested_params$wrap_nested_params$fn__24688.invoke(nested_params.clj:89)
	at ring.middleware.multipart_params$wrap_multipart_params$fn__24986.invoke(multipart_params.clj:173)
	at ring.middleware.params$wrap_params$fn__25010.invoke(params.clj:67)
	at ring.middleware.cookies$wrap_cookies$fn__24467.invoke(cookies.clj:175)
	at ring.middleware.absolute_redirects$wrap_absolute_redirects$fn__25116.invoke(absolute_redirects.clj:47)
	at ring.middleware.resource$wrap_resource$fn__25026.invoke(resource.clj:37)
	at ring.middleware.content_type$wrap_content_type$fn__25064.invoke(content_type.clj:34)
	at ring.middleware.default_charset$wrap_default_charset$fn__25088.invoke(default_charset.clj:31)
	at ring.middleware.not_modified$wrap_not_modified$fn__17997.invoke(not_modified.clj:53)
	at ring.middleware.x_headers$wrap_x_header$fn__23471.invoke(x_headers.clj:22)
	at ring.middleware.x_headers$wrap_x_header$fn__23471.invoke(x_headers.clj:22)
	at ring.middleware.x_headers$wrap_x_header$fn__23471.invoke(x_headers.clj:22)
```

When reaching the step of the callback URI.

I suspect associng the `access-token` to the session is the culprit.

### PR DESCRIPTION
**Pull request by Heliosmaster**
The newer version of clj-time supports tagged literals for clj-time/date-time. Fixes #27

### CODE REVIEW

**Comment by weavejester** [comment]
Thanks for the PR. Can you add a small test to ensure that the middleware works with Ring's cookie sessions store?

**Comment by Heliosmaster** [comment]
Hey @weavejester I have reworked this according to your comments on #27 . Now the expiration is stored internally as a `java.util.Date` which should be a bit more robust. It adds a bit of code to do the conversion back and forth. I have also taken the liberty of deleting a couple of unused expression in tests.

**Comment by mtravers** [comment]
Any reason this has been in limbo for so long? This problem is biting me and I guess I will make my own fork.

**Comment by weavejester** [comment]
Occassionally a PR will slip through the net. It's why in the contributing guidelines I added a line saying that it's okay to pester me. Overall the PR looks fine, it just needs some formatting updates to fit in with the style guidelines.

**Comment by Heliosmaster** [comment]
@weavejester I've merged master in, could you point where formatting is not up to standard? It should follow the normal clojure style guide already 🤔

**Comment by weavejester** [comment]
There are a few lines that exceed 80 characters that need to be fixed, but I think that's the only style issue. However, there look to be some lines in the test code already that hit 85, so as long as the lines are 80-ish it should be okay, and we can be stricter in a future formatting pass.

The ending newline of the file has been accidentally removed and should be put back. Also since the other `let` clauses are aligned, the one this patch adds should also have its symbols/values aligned into columns.

**Comment by malyn** [comment]
I need this fix as well and would love to push @Heliosmaster's PR over the finish line. I don't think that there is a way for me to add my own commit to the PR, so I took the PR branch, rebased to master, and then [added a commit](https://github.com/malyn/ring-oauth2/commit/49e0f25ede8821cc6500ff9a930788f5a3228ae9) that addresses the formatting issues. The resulting branch is here: <https://github.com/malyn/ring-oauth2/tree/patch-1>

@weavejester Let me know if this addresses your formatting comments, and if so, how to get this extra commit merged into the PR. I can submit a PR against @Heliosmaster's patch branch if that helps, for example.

And just to be clear, this is all @Heliosmaster's great work in addressing the underlying issue! I just made some style edits.

**Comment by weavejester** [comment]
Your patch has some non-standard formatting, @malyn. I follow the Clojure Style Guide for my projects, so the indentation for macros like `->` should be consistent.

I think we can also do something like:

```clojure
(defn- coerce-token-expires-date [tokens]
  (into {} (for [[id t]] [id (update t :expires coerce/from-date)])))
```

**Comment by malyn** [comment]
Thanks, @weavejester, for the quick response! I [factored out](https://github.com/malyn/ring-oauth2/commit/9cd1ac87192d11fe87fb9290571b51f5bfec24a7) the `coerce-tokens-expires-date` function and I agree that looks a lot better.

I really struggled with how to handle `->` in this patch. Normally I would have vertically aligned that (per the Clojure Style Guide) and that is what I do in my own code So [these lines from the patch](https://github.com/malyn/ring-oauth2/blob/49e0f25ede8821cc6500ff9a930788f5a3228ae9/test/ring/middleware/oauth2_test.clj#L85-L86) would become this:

```clojure
(-> response
    :session
    ::oauth2/access-tokens
    :test
    :expires
    time-coerce/from-date)
```

I didn't see that vertical form used *on keywords* anywhere else in the file though. Instead, I saw the indentation that I ultimately ended up using (from [these lines in master](https://github.com/weavejester/ring-oauth2/blob/c30c6990e0558d5bdfab21a6f0af539797f92144/test/ring/middleware/oauth2_test.clj#L205-L206), for example). I am happy to change the patch to use the vertical form, I was just trying to follow the existing code as closely as possible.

Please let me know how you would like to proceed and I will make the requisite change.

**Comment by weavejester** [comment]
This would be fine:

```clojure
(-> response :session ::oauth2/access-tokens
    :test :expires time-coerce/from-date)
```

Or however you want to break it up.

**Comment by weavejester** [comment]
Also the `time-coerce` alias could reasonably just be `coerce` for brevity.

**Comment by malyn** [comment]
Very good, thank you for that guidance, @weavejester! I changed `time-coerce` to `coerce` and then reformatted the code according to your first proposal. I felt that this better grouped the threading phrase into two sections: get the access tokens, then pull a specific thing out of one of those tokens. This matched what was in nearby lines as well. I also took the liberty of reformatting a few of those nearby lines to get them under 80 characters in width, and to make everything look visually consistent in these updated code blocks.

I recommend you look at the full diff against master (instead of just the recent commit) to get a sense of the final sense of the final patch: <https://github.com/weavejester/ring-oauth2/compare/master...malyn:patch-1>

Let me know if you have any other changes. If not, I can submit a PR for these style changes to @Heliosmaster's original branch, which, if accepted, *should* automatically update this PR with these style changes. I have never done this before though, so we may have to take another approach if that fails somewhere along the chain.

**Comment by Heliosmaster** [comment]
Sorry, I totally missed this. Thanks @malyn! Submit the PR and i'll accept it and this should be updated too 👏

**Comment by Heliosmaster** [comment]
I've merged @malyn 's branch `patch-1` into my `patch-1`

**Comment by jeroenvandijk** [comment]
FYI, I had the same issue and this branch has fixed it. I've created a [fork](https://github.com/jeroenvandijk/ring-oauth2/tree/deps.edn) with this branch plus a `deps.edn` so people can try it out in their own deps project. 
```
ring/ring-oauth {:git/url "https://github.com/jeroenvandijk/ring-oauth2.git" :sha "9af66afc6827db18d17eef4cdf64532c9d351cce"}
```

**Comment by lockie** [comment]
@weavejester Hey, do you mind merging this PR for those of us who use lein and can't install from github forks?

File: src/ring/middleware/oauth2.clj
Comment (weavejester) [inline src/ring/middleware/oauth2.clj:113]: Might be better to use:

```clojure
(persistent!
 (reduce-kv #(assoc %1 %2 (update %3 :expires coerce/from-date))
            (transient {})
            tokens)
```

That way we're not creating an intermediate seq, but instead adding directly to a transient map.

**Review by weavejester — CHANGES_REQUESTED** [review]
Looks good; some minor changes necessary before it can be merged, and the commits will need to be squashed down into a clean history.

**Comment by gerdint** [comment]
So has this PR stalled? This issue is biting me.

**Comment by weavejester** [comment]
The PR is waiting on the author to address the findings from the code review. I can take a look at addressing those myself (or someone else can), but typically I remove PRs that I've reviewed from my todo list until I get a response.

**Comment by gerdint** [comment]
Right, well it's been more than a year so I guess @Heliosmaster is up to other things. But I'm using deps so I'll just try the fork for now. But should there not be any major issues it would be nice to get this merged and not have to put up with JodaTime.

**Comment by gerdint** [comment]
Since clj-time is deprecated it would also be good to migrate away from it I guess, provided Java 8+ is OK. I noticed that I still had issues with malli complaining even with jeroenvandijk's branch (though I checked that date serialization and deserialization looked ok, so it must be something else).

**Pull request closed without merging**
